File: 2022/22/solution.hs
import Data.Char
import Data.List
import Data.List.Split (splitOn)
import qualified Data.Map as Map
import Data.Maybe

data Instruction
  = L
  | R
  | M Int
  deriving (Show)

data Tile
  = GROUND
  | STONE
  | EMPTY
  deriving (Show, Eq)

data Coord =
  C Int Int
  deriving (Eq, Ord, Show)

directions = [C 0 1, C 1 0, C 0 (-1), C (-1) 0]
numDirections = 4

getX (C x _) = x

getY (C _ y) = y

addCoord (C x1 y1) (C x2 y2) = C (x1 + x2) (y1 + y2)

data CubeFace =
  CubeFace
    { value :: Int
    , rotation :: Int
    }
  deriving (Eq, Ord, Show)

rotateCubeFace (CubeFace v r) rot = CubeFace v (mod (r + rot) numDirections)

data GridMap =
  GridMap
    { grid :: [[Tile]]
    , numRows :: Int
    , numCols :: Int
    , blockSize :: Int
    , instructions :: [Instruction]
    , initCoord :: Coord
    }

parse :: String -> GridMap
parse s = gridMap
  where
    sl = splitOn "\n\n" s
    numCols = maximum $ map length unpaddedGrid
    numRows = length unpaddedGrid
    numNoneGround = length $ filter (/= EMPTY) $ concat grid
    blockSize = (floor . sqrt . fromIntegral) (numNoneGround `div` 6)

    gridMap = GridMap grid numRows numCols blockSize instructions initCoord
    initCoord = C 0 (fromJust (elemIndex GROUND (head grid)))
    grid = map (\r -> take numCols (r ++ repeat EMPTY)) unpaddedGrid

    unpaddedGrid = map (map gridHelper) $ lines $ head sl
    gridHelper ' ' = EMPTY
    gridHelper '.' = GROUND
    gridHelper '#' = STONE

    instructions = helper (sl !! 1)
    helper (c:cs)
      | c == 'L' = L : helper cs
      | c == 'R' = R : helper cs
      | otherwise = M (read (takeWhile isDigit (c : cs))) : helper (dropWhile isDigit cs)
    helper [] = []

evaluate findNext gridMap = (finalX + 1) * 1000 + numDirections * (finalY + 1) + finalD
  where
    (GridMap grid numRows numCols _ instructions initCoord) = gridMap
    (C finalX finalY, finalD) = foldl move (initCoord, 0) instructions
    move (initCoord, dindex) L = (initCoord, mod (dindex - 1) numDirections)
    move (initCoord, dindex) R = (initCoord, mod (dindex + 1) numDirections)
    move (initCoord, dindex) (M n) = nextGridPoint
      where
        (nextCoord, nextDindex) = findNext (initCoord, dindex)
        gridPoint = (grid !! getX nextCoord) !! getY nextCoord

        nextGridPoint
          | n == 0 = (initCoord, dindex)
          | gridPoint == STONE = (initCoord, dindex)
          | gridPoint == GROUND = move (nextCoord, nextDindex) (M (n - 1))
          | gridPoint == EMPTY = error "Cannot travel to an empty square."

part1 :: GridMap -> Int
part1 gridMap = evaluate findNext gridMap
  where
    (GridMap grid numRows numCols _ instructions initCoord) = gridMap
    findNext (initCoord, dindex) = (nextCoord, nextDindex)
      where
        direction = directions !! dindex
        nextCoord =
          head $
          dropWhile (\(C r c) -> ((grid !! r) !! c) == EMPTY) $
          tail $ 
          iterate (\(C r c) -> C (mod (r + getX direction) numRows) (mod (c + getY direction) numCols)) initCoord
        nextDindex = dindex

cubeEdgesHalf =
  [ (CubeFace 1 0, CubeFace 2 3)
  , (CubeFace 1 1, CubeFace 4 2)
  , (CubeFace 1 2, CubeFace 5 3)
  , (CubeFace 2 0, CubeFace 3 3)
  , (CubeFace 2 1, CubeFace 6 2)
  , (CubeFace 3 0, CubeFace 1 3)
  , (CubeFace 3 1, CubeFace 5 2)
  , (CubeFace 3 2, CubeFace 6 3)
  , (CubeFace 4 0, CubeFace 6 1)
  , (CubeFace 4 3, CubeFace 2 2)
  , (CubeFace 5 0, CubeFace 4 1)
  , (CubeFace 5 1, CubeFace 6 0)
  ]

cubeEdges = Map.fromList $ cubeEdgesHalf ++ map (\v -> (snd v, fst v)) cubeEdgesHalf

foldCube (GridMap grid numRows numCols blockSize _ _) = assignFaces blocksMapInit [firstBlock]
  where
    numRowBlocks = div numRows blockSize
    numColBlocks = div numCols blockSize
    blocks =
      [ C (div r blockSize) (div c blockSize)
      | r <- [0,blockSize .. numRows - 1]
      , c <- [0,blockSize .. numCols - 1]
      , (grid !! r) !! c /= EMPTY
      ]
    firstBlock = head blocks
    blocksMapInit = Map.fromList [(firstBlock, CubeFace 1 0)]
    assignFaces blocksMap nextSeq
      | length blocksMap == length blocks = blocksMap
      | otherwise                         = assignFaces updatedBlocksMap updatedNextSeq
      where
        currentBlock = head nextSeq
        CubeFace cv cr = fromJust $ Map.lookup currentBlock blocksMap
        nFacesDir =
          map rotationHelper $
          filter (\(c, _) -> elem c blocks && notElem c nextSeq && Map.notMember c blocksMap) $
          map (\d -> (addCoord currentBlock (directions !! d), d)) [0 .. 3]

        rotationHelper (C r c, rot) = (C r c, assignment)
          where
            key = CubeFace cv (mod (cr + rot) numDirections)
            (CubeFace av ar) = fromJust (Map.lookup key cubeEdges)
            assignment = CubeFace av (mod (ar - rot + 2) numDirections)

        updatedBlocksMap = Map.union blocksMap (Map.fromList nFacesDir)
        updatedNextSeq = tail nextSeq ++ map fst nFacesDir


part2 :: GridMap -> Int
part2 gridMap = evaluate findNext gridMap
  where
    (GridMap grid numRows numCols blockSize _ _) = gridMap
    blockAssignments = foldCube gridMap
    invBlockAssignments = Map.fromList $ map (\(k, v) -> (value v, k)) $ Map.toList blockAssignments

    findNext (initCoord, dindex) = (nextCoord, nextDindex)
      where
        (C ir ic) = initCoord
        direction = directions !! dindex
        candCoord = C (ir + getX direction) (ic + getY direction)
        (C rc cc) = candCoord
        currentBlock = C (div ir blockSize) (div ic blockSize)
        candBlock = C (div rc blockSize) (div cc blockSize)

        exitEdge = rotateCubeFace (fromJust (Map.lookup currentBlock blockAssignments)) dindex
        nextEdge = fromJust (Map.lookup exitEdge cubeEdges)
        nextBlock = fromJust (Map.lookup (value nextEdge) invBlockAssignments)
        nextAssignment = fromJust (Map.lookup nextBlock blockAssignments)
        nextDindex
          | currentBlock == candBlock = dindex
          | otherwise                 = mod (rotation nextEdge - rotation nextAssignment + 2) numDirections
        blockCoord = C (ir - getX currentBlock * blockSize) (ic - getY currentBlock * blockSize)
        exitCoord
          | dindex == 0 = getX blockCoord
          | dindex == 1 = blockSize - 1 - getY blockCoord
          | dindex == 2 = blockSize - 1 - getX blockCoord
          | otherwise   = getY blockCoord
        enterCoord
          | nextDindex == 0 = C exitCoord 0
          | nextDindex == 1 = C 0 (blockSize - 1 - exitCoord)
          | nextDindex == 2 = C (blockSize - 1 - exitCoord) (blockSize - 1)
          | otherwise = C (blockSize - 1) exitCoord
        nextCoord
          | currentBlock == candBlock = candCoord
          | otherwise = C (getX enterCoord + getX nextBlock * blockSize) (getY enterCoord + getY nextBlock * blockSize)

main = do
  testInputString <- readFile "test_input.dat"
  let testGridMap = parse testInputString
  inputString <- readFile "input.dat"
  let gridMap = parse inputString
  print (part1 testGridMap)
  print (part1 gridMap)
  print (part2 testGridMap)
  print (part2 gridMap)

File: 2022/22/solution.py
import itertools
import math
import typing

DIRECTIONS = [(0, 1), (1, 0), (0, -1), (-1, 0)]
NUM_DIRECTIONS = 4


def parse(path):
    with open(path) as f:
        txt = f.read().split("\n\n")
        instructions = txt[1].strip()
        grid = txt[0].splitlines()
        max_length = max(len(l) for l in grid)
        grid = [l + " " * (max_length - len(l)) for l in grid]

        instructions = [
            "".join(v[1])
            for v in itertools.groupby(instructions, key=lambda c: c.isdigit())
        ]
        instructions = [int(v) if v[0].isdigit() else v for v in instructions]
    return instructions, grid


def part1(instructions, grid):
    num_rows = len(grid)
    num_col = len(grid[0])
    coord = [0, grid[0].index(".")]
    dindex = 0
    direction = DIRECTIONS[dindex]
    for instr in instructions:
        if instr == "R":
            dindex = (dindex + 1) % NUM_DIRECTIONS
            direction = DIRECTIONS[dindex]
        elif instr == "L":
            dindex = (dindex - 1) % NUM_DIRECTIONS
            direction = DIRECTIONS[dindex]
        else:
            for _ in range(instr):
                next_coord = list(coord)
                next_coord[0] = (next_coord[0] + direction[0]) % num_rows
                next_coord[1] = (next_coord[1] + direction[1]) % num_col

                if grid[next_coord[0]][next_coord[1]] == ".":
                    coord = next_coord
                elif grid[next_coord[0]][next_coord[1]] == "#":
                    break
                else:
                    while grid[next_coord[0]][next_coord[1]] == " ":
                        next_coord[0] = (next_coord[0] + direction[0]) % num_rows
                        next_coord[1] = (next_coord[1] + direction[1]) % num_col
                    if grid[next_coord[0]][next_coord[1]] == ".":
                        coord = next_coord
                    elif grid[next_coord[0]][next_coord[1]] == "#":
                        break

    print(coord)
    return (coord[0] + 1) * 1000 + 4 * (coord[1] + 1) + dindex


class CubeFace(typing.NamedTuple):
    value: int
    rotation: int


# Define how the edges of a right handed dice touch, and under which angle/rotation.
CUBE_EDGES = {
    CubeFace(value=1, rotation=0): CubeFace(value=2, rotation=3),
    CubeFace(value=1, rotation=1): CubeFace(value=4, rotation=2),
    CubeFace(value=1, rotation=2): CubeFace(value=5, rotation=3),
    CubeFace(value=2, rotation=0): CubeFace(value=3, rotation=3),
    CubeFace(value=2, rotation=1): CubeFace(value=6, rotation=2),
    CubeFace(value=3, rotation=0): CubeFace(value=1, rotation=3),
    CubeFace(value=3, rotation=1): CubeFace(value=5, rotation=2),
    CubeFace(value=3, rotation=2): CubeFace(value=6, rotation=3),
    CubeFace(value=4, rotation=0): CubeFace(value=6, rotation=1),
    CubeFace(value=4, rotation=3): CubeFace(value=2, rotation=2),
    CubeFace(value=5, rotation=0): CubeFace(value=4, rotation=1),
    CubeFace(value=5, rotation=1): CubeFace(value=6, rotation=0),
}
CUBE_EDGES = CUBE_EDGES | dict((v, k) for k, v in CUBE_EDGES.items())


def fold_cube(grid):
    """BFS to figure out which block corresponds to which number and rotation of the cube."""
    grid_res = int(math.sqrt(sum([c != " " for l in grid for c in l]) // 6))
    n = len(grid) // grid_res
    m = len(grid[0]) // grid_res
    blocks = [
        (i, j)
        for i in range(n)
        for j in range(m)
        if grid[i * grid_res][j * grid_res] != " "
    ]

    block_assignments = {blocks[0]: CubeFace(1, 0)}
    q = [blocks[0]]

    while q:
        block = q.pop(0)
        face = block_assignments[block]
        (block_i, block_j) = block
        for epi, (delta_i, delta_j) in enumerate(DIRECTIONS):
            next_block = (block_i + delta_i, block_j + delta_j)
            if next_block not in blocks or next_block in block_assignments:
                continue

            ni, nj = next_block
            if not ((0 <= ni < n) and (0 <= nj < m)):
                continue

            active_edge = CubeFace(face.value, (epi + face.rotation) % NUM_DIRECTIONS)
            nm = CUBE_EDGES[active_edge]
            r = (nm.rotation - epi + 2) % NUM_DIRECTIONS

            block_assignments[next_block] = CubeFace(nm.value, r)
            q.append(next_block)

    inv_block_assignments = {v.value: k for k, v in block_assignments.items()}
    return block_assignments, inv_block_assignments, grid_res


def part2(instructions, grid):
    block_assignments, inv_block_assignments, grid_res = fold_cube(grid)
    blocks = sorted(block_assignments)
    current_block = blocks[0]
    coord, dindex = [current_block[0] * grid_res, current_block[1] * grid_res], 0

    for instr in instructions:
        if instr == "R":
            dindex = (dindex + 1) % NUM_DIRECTIONS
        elif instr == "L":
            dindex = (dindex - 1) % NUM_DIRECTIONS
        else:
            for _ in range(instr):
                direction = DIRECTIONS[dindex]
                next_coord = list(coord)
                next_coord[0] = next_coord[0] + direction[0]
                next_coord[1] = next_coord[1] + direction[1]
                next_block = (next_coord[0] // grid_res, next_coord[1] // grid_res)
                next_dindex = dindex

                if next_block != current_block:
                    current_assignment = block_assignments[current_block]

                    exit_edge = CubeFace(
                        current_assignment.value,
                        (current_assignment.rotation + dindex) % NUM_DIRECTIONS,
                    )
                    next_edge = CUBE_EDGES[exit_edge]

                    next_block = inv_block_assignments[next_edge.value]
                    next_assignment = block_assignments[next_block]

                    next_dindex = (
                        next_edge.rotation - next_assignment.rotation + 2
                    ) % NUM_DIRECTIONS
                    block_coord = [
                        coord[0] - current_block[0] * grid_res,
                        coord[1] - current_block[1] * grid_res,
                    ]

                    if dindex == 0:
                        rel_coord = block_coord[0]
                    elif dindex == 1:
                        rel_coord = grid_res - 1 - block_coord[1]
                    elif dindex == 2:
                        rel_coord = grid_res - 1 - block_coord[0]
                    else:
                        rel_coord = block_coord[1]

                    if next_dindex == 0:
                        next_coord = (rel_coord, 0)
                    elif next_dindex == 1:
                        next_coord = (0, grid_res - 1 - rel_coord)
                    elif next_dindex == 2:
                        next_coord = (grid_res - 1 - rel_coord, grid_res - 1)
                    else:
                        next_coord = (grid_res - 1, rel_coord)

                    next_coord = [
                        next_block[0] * grid_res + next_coord[0],
                        next_block[1] * grid_res + next_coord[1],
                    ]

                if grid[next_coord[0]][next_coord[1]] == ".":
                    coord = next_coord
                    current_block = next_block
                    dindex = next_dindex
                elif grid[next_coord[0]][next_coord[1]] == "#":
                    break
                else:
                    raise Exception("Something unexpected happened!")

    return 1000 * (coord[0] + 1) + 4 * (coord[1] + 1) + dindex


def main():
    test_instructions, test_grid = parse("test_input.dat")
    instructions, grid = parse("input.dat")
    print(fold_cube(grid))

    assert part1(test_instructions, test_grid) == 6032
    print(part1(instructions, grid))

    assert part2(test_instructions, test_grid) == 5031
    print(part2(instructions, grid))


if __name__ == "__main__":
    main()
